src/System.Xml.XmlSerializer/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs

```diff
@@ -0,0 +1,1242 @@
+            {
+                if (member.Collection != null)
+                {
+                    MemberInfo[] memberInfos = o.GetType().GetMember(member.Mapping.Name);
+                    var memberInfo = memberInfos[0];
+                    object collection = null;
+                    SetCollectionObjectWithCollectionMember(ref collection, member.Collection, member.Mapping.TypeDesc.Type);
+                    SetMemberValue(o, collection, memberInfo);
+                }
+            }
+        }
+
+        private void SetCollectionObjectWithCollectionMember(ref object collection, CollectionMember collectionMember, Type collectionType)
+        {
+            if (collectionType.IsArray)
+            {
+                Array a;
+                Type elementType = collectionType.GetElementType();
+                var currentArray = collection as Array;
+                if (currentArray != null && currentArray.Length == collectionMember.Count)
+                {
+                    a = currentArray;
+                }
+                else
+                {
+                    a = Array.CreateInstance(elementType, collectionMember.Count);
+                }
+
+                for (int i = 0; i < collectionMember.Count; i++)
+                {
+                    a.SetValue(collectionMember[i], i);
+                }
+
+                collection = a;
+            }
+            else
+            {
+                if (collection == null)
+                {
+                    collection = ReflectionCreateObject(collectionType);
+                }
+
+                if (typeof(IList).IsAssignableFrom(collectionType))
+                {
+                    foreach (var item in collectionMember)
+                    {
+                        ((IList)collection).Add(item);
+                    }
+                }
+                else
+                {
+                    foreach (var item in collectionMember)
+                    {
+                        MethodInfo addMethod = collectionType.GetMethod("Add");
+                        if (addMethod == null)
+                        {
+                            throw new InvalidOperationException("(addMethod == null)");
+                        }
+
+                        addMethod.Invoke(collection, new object[] { item });
+                    }
+                }
+            }
+        }
+
+        private static void SetMemberValue(object o, object value, string memberName)
+        {
+            MemberInfo[] memberInfos = o.GetType().GetMember(memberName);
+            var memberInfo = memberInfos[0];
+            SetMemberValue(o, value, memberInfo);
+        }
+
+        private static void SetMemberValue(object o, object value, MemberInfo memberInfo)
+        {
+            if (memberInfo is PropertyInfo)
+            {
+                var propInfo = (PropertyInfo)memberInfo;
+                propInfo.SetValue(o, value);
```

Comment:
I presume this is a "get it working" step and you will optimize things like delegate setters later?